crates/utils/re_capabilities/src/main_thread_token.rs

```diff
@@ -0,0 +1,53 @@
+use static_assertions::assert_not_impl_any;
+
+/// A token that proves we are on the main thread.
+///
+/// Certain operations are only allowed on the main thread.
+/// These operations should require this token.
+/// For instance, any function using file dialogs (e.g. using [`rfd`](https://docs.rs/rfd/latest/rfd/)) should require this token.
+///
+/// The token should only be constructed in `fn main`, using [`MainThreadToken::i_promise_i_am_on_the_main_thread`],
+/// and then be passed down the call tree to where it is needed.
+/// [`MainTheadToken`] is neither `Send` nor `Sync`,
+/// thus guaranteeing that it cannot be found in other threads.
+///
+/// Of course, there is nothing stopping you from calling [`MainThreadToken::i_promise_i_am_on_the_main_thread`] from a background thread,
+/// but PLEASE DON'T DO THAT.
+/// In other words, don't use this as a guarantee for unsafe code.
+///
+/// There is also [`MainThreadToken::from_egui_ui`] which uses the implicit guarantee of egui
+/// (which _usually_ is run on the main thread) to construct a [`MainThreadToken`].
+/// Use this only in a code base where you are sure that egui is running only on the main thread.
+#[derive(Clone, Copy)]
+pub struct MainThreadToken {
+    /// Prevent from being sent between threads.
+    ///
+    /// Workaround until `impl !Send for X {}` is stable.
+    _dont_send_me: std::marker::PhantomData<*const ()>,
+}
+
+impl MainThreadToken {
+    /// Only call this from `fn main`, or you may get weird runtime errors!
+    pub fn i_promise_i_am_on_the_main_thread() -> Self {
+        #[cfg(not(target_arch = "wasm32"))]
+        debug_assert_eq!(std::thread::current().name(), Some("main"),
+            "DEBUG ASSERT: Trying to construct a MainThreadToken on a thread that is not the main thread!"
+        );
+
+        Self {
+            _dont_send_me: std::marker::PhantomData,
+        }
+    }
+
+    /// We should only create an `egui::Ui` on the main thread,
+    /// so having it is proof enough that we are on the main thread.
+    ///
+    /// Use this only in a code base where you are sure that egui is running only on the main thread.
+    #[cfg(feature = "egui")]
+    pub fn from_egui_ui(_ui: &egui::Ui) -> Self {
```

Comment:
It seems to me like you can trivially create a `egui::Ui` on any thread:

```rust
egui::Ui::new(egui::Context::default(), egui::Id::new("test"), egui::UiBuilder::default());
```

And there isn't really any guarantee that someone won't do that.